### PR TITLE
adds composer.json file with wordpress dependecies, updates .gitignor…

### PR DESCRIPTION
**Pull request by hunterlester**
@ericelliott 
See how this works for you.

`vagrant ssh` && `cd platform`

`curl -sS https://getcomposer.org/installer | php` to install Composer
(I'm assuming that Composer can potentially be added to the Docker image for installation)

The terminal may ask you to `sudo apt-get install php5-cli`

run `php composer.phar install` to install plugins

Although I'm making a pull request, I don't mean for this fork to be merged, I just wanted you to check it out.

### CODE REVIEW

**Comment by hunterlester** [comment]
I wanted to install Composer in the root directory but for some reason it's not so straight forward as to how to redirect the plugins to `/platform/wp-content/plugins`, so for now composer.json is in `/platform`

**Comment by ericelliott** [comment]
Cool. You should be able to edit the machine configuration to do the install process automatically. =)

**Comment by hunterlester** [comment]
In the Vagrantfile or is there a separate provision file that I've overlooked?

**Comment by ericelliott** [comment]
@mattma What's the right place to install things on the container?

**Comment by ericelliott** [comment]
Very cool.. pinging @mattma &/or @robmazan for quick review.

File: platform/composer.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "jshomes/wordpress-platform",
+    "description": "Plugin management for JS Homes Platform",
+    "require": {
+        "wpackagist-plugin/twitter": "1.4.0",
+        "wpackagist-plugin/affiliates-manager": "2.1.7",
+        "wpackagist-plugin/sendgrid-email-delivery-simplified": "1.6.7",
+        "wpackagist-plugin/profile-builder": "2.2.6",
+        "wpackagist-plugin/usersnap": "4.3",
+        "wpackagist-plugin/wp-super-cache": "1.4.6",
```
Comment (ericelliott) [inline platform/composer.json]: Oops. I'm not using this anymore. I've replaced it with the WordFence cache. Do I have a ticket for the WordFence security plugin?

**Comment by hunterlester** [comment]
I'll be watching plugins over next several days to see if they update automatically. Wpackagist docs state that they update their repos every hour or so, watching for Wordpress plugin updates. Let's see..

**Comment by ericelliott** [comment]
Very cool! Thanks a lot for taking a crack at this. I'll look at it as soon as I have a moment.

**Comment by hunterlester** [comment]
Really exciting for me to be a part of this!! :-D

**Comment by robmazan** [comment]
> What's the right place to install things on the container?

Hi @ericelliott, @hunterlester,

If you want to install Composer on the Wordpress container you better add those commands to this block in the `Dockerfile`: https://github.com/jshomes/JSHomes-Platform/blob/master/platform/Dockerfile#L6-L10

The commands in the `Vagrantfile` are running on the VM that is running the containers, not on the container itself. You have three participants here:

1. Your host computer running VirtualBox
2. Your VM ran by your VirtualBox, running docker
3. The wordpress container ran by docker

These are three different file systems, and what can be confusing is that the project folder is shared twice here: it is shared from your host to the VirtualBox VM, and it is shared further from the VirtualBox VM to the container (at least some parts of it).

So basically anything you want to run on the wordpress container must be in the `Dockerfile`. If you need some files to be present on the container when you run those commands (like `composer.json`), you can put that inside the `platform` folder, and use the `ADD` command in the `Dockerfile` to have docker copy over that file to the container's file system. (http://docs.docker.com/engine/reference/builder/#add)

**Comment by hunterlester** [comment]
@robmazan Thank you for taking the time to explain that and giving me a little education :-D

**Comment by hunterlester** [comment]
PHP5 is installing, I just have to figure out how to properly install Composer.
I'll have to get back to it tomorrow.

**Comment by ericelliott** [comment]
(☞ﾟ∀ﾟ)☞ @robmazan rocks! 🎸

**Comment by hunterlester** [comment]
@ericelliott @robmazan  Ok, finally got this working. Necessary commands are scripted in Vagrantfile. I tested it and the plugins are installing automatically after running `vagrant up`.

I'm going to have to dig into Docker tutorials and docs because I'm not understanding how it works.

**Comment by ericelliott** [comment]
Rad. =)

@robmazan got time for one more peek?

**Comment by hunterlester** [comment]
I suppose the next thing I need to do is write in proper version constraints for each plugin.

I've got some reading to do on that. Learning a lot here!! :-D

**Comment by ericelliott** [comment]
Great! Thanks a million for the help, @hunterlester (☞ﾟ∀ﾟ)☞ you rock!

**Comment by robmazan** [comment]
Hi @ericelliott, @hunterlester,

Can't check it out right now, but based on the sources I would suggest that rather than having Composer installed on the Vagrant VM, it should be used from a docker container. That can be either the existing WordPress container, or a dedicated Composer instance.

The point of the Vagrant VM is only to provide tooling that makes possible running docker containers, so I would keep that single responsibility for that VM, and use containerized apps for everything else.

So I would either add the Composer installation to the Dockerfile like:

```Dockerfile
ADD composer.json /var/www/html
RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer && \
    composer install
```

(Not sure if I'm using the correct paths here, but the first command is copying over the `composer.json` from your platform directory to the docker container's file system, and the second one is installing and running composer.)

The other solution is to use a dedicated container for installing dependencies with Composer as described here: https://hub.docker.com/r/composer/composer/

The latter might be even cleaner, as in that case you leave your WordPress container for WordPress only, and not piling up additional responsibilities for that container.

**Comment by hunterlester** [comment]
@robmazan Thank you. I'll dig into working on that.

**Comment by hunterlester** [comment]
Be back soon. These next couple of weeks will have me jamming to finish a client app.
I'll schedule an hour or two each day to work on this fork.

**Comment by ericelliott** [comment]
Cool. =)

**Comment by mattma** [comment]
@ericelliott I have been busy with other projects lately and tied up most of my time. @hunterlester how is the status of the PR? Do you finish everything on your end?

**Comment by hunterlester** [comment]
@mattma Good morning Matt. I haven't finished yet. I need Composer to be in it's own Docker container and I'm figuring out how to do that. Working my way through docs and tutorials.

**Comment by hunterlester** [comment]
@ericelliott @mattma @robmazan The composer.json and composer.lock files are being `ADD`ed  and Composer image is being built. The last thing I need to do is to get Composer to install the dependencies  upon `start` up of the Docker containers. This command `docker run -v $(pwd):/app composer/composer install` is cited in the DockerHub Composer documentation but I'm not sure how to `RUN` it in the dockerfile. 

I've experimented with `docker run -i -t platform_wordpress /bin/bash` to manually run  `docker run -v $(pwd):/app composer/composer install`

What confuses me is that the `-v` flag is used to mount a directory from the host (Vagrant) into the container, however, my goal is to just get the Composer container to install dependencies based on the json files which have already been `ADD`ed to the `platform_wordpress` container.

**Comment by robmazan** [comment]
For that setup I think it would be easier to move the `composer.json` and `composer.lock` files to the folder where you want to install the plugins (`wp-content`?), and instead of adding composer container to the `docker-compose.yml`, add the command form to the shell provisioning part of the [`Vagrantfile`](https://github.com/hunterlester/JSHomes-Platform/blob/master/Vagrantfile#L81), having docker to mount the `wp-content` folder as a volume, like:

```sh
docker run -v /vagrant/platform/wp-content:/app composer/composer install
```

That will make available the `wp-content` folder to the composer container under `/app`, where it will execute a `composer install` command, then exit. Even though that happens in the container, due to the volume mounting, you'll have your dependencies installed to your actual `wp-content` folder.

Does that meet your goals?

**Comment by hunterlester** [comment]
@robmazan @ericelliott This installed the plugins into the host directory but not the container. Nonetheless I'll keep trying to get this working. I've opened up a stack overflow question.

**Comment by robmazan** [comment]
@hunterlester 

Where does Composer install the files? What's the exact, project relative path? Maybe it is not mounted as a volume on the container.

**Comment by hunterlester** [comment]
@robmazan In this last case, Composer installed the `/vendor` directory in `/home/vagrant/platform` and the rest of the plugins in `/home/vagrant/platform/wp-content/plugins`

Project relative path in the container? `/var/www/html/`

**Comment by robmazan** [comment]
@hunterlester based on that, the stuff under `plugins` should be available on the container, but the `vendor` not. Only the `plugins` and `themes` folders are configured as volumes for the container: https://github.com/jshomes/JSHomes-Platform/blob/master/platform/docker-compose.yml.tmpl#L6-L7

You might need to add the `vendor` folder to that list. For convenience, add it to the [Dockerfile](https://github.com/jshomes/JSHomes-Platform/blob/master/platform/Dockerfile#L20) as well.

If still not working, please push the changes to your branch, I might have some time now to check out what actually happens.

**Comment by hunterlester** [comment]
@robmazan Cool. I'll do that right now.

**Comment by hunterlester** [comment]
@robmazan @ericelliott 
That didn't install the plugins into the container's working directory, still in the Vagrant directory.

Here's my question, to reiterate what I'm trying to do: http://stackoverflow.com/questions/34212477/docker-how-to-install-dependencies-using-separate-composer-container-in-word

To be sure that I'm understanding how you want this setup, is the goal to just have the plugins be installed in wordpress container?

**Comment by robmazan** [comment]
Yep, as in the end it will be needed by the running WordPress, and that's running in the container.

For the changes:

Here you are missing a leading slash from the path you've added: https://github.com/hunterlester/JSHomes-Platform/commit/b3edb42975f759901d4b9684f762ffd07a2356bf?w=1#diff-6c7cfa47c55f9ab986164c0e2524c187R25

Instead of [adding](https://github.com/jshomes/JSHomes-Platform/pull/45/files?w=1#diff-6c7cfa47c55f9ab986164c0e2524c187R5) `composer.json`, just put it under the `wp-content` folder. It will be there for the composer container, if you use that path with the docker command in your `Vagrantfile`:

```sh
docker run -v /vagrant/platform/wp-content:/app composer/composer install
```

And finally, don't add composer to the `links` part in the [`docker-composer.yml`](https://github.com/jshomes/JSHomes-Platform/pull/45/files?w=1#diff-f37bdf89eb683313f7ba67356994ba61R5), rather add `vendor` to the volumes as:

```
   volumes:
     - wp-content/plugins/:/var/www/html/wp-content/plugins
     - wp-content/themes/:/var/www/html/wp-content/themes
     - wp-content/vendor/:/var/www/html/wp-content/vendor
```

**Comment by hunterlester** [comment]
@robmazan 
`/vendor` is generated when Composer runs to install dependencies. '/vendor' is not present in host directory before install happens.

Here's what I'm thinking:
- `ADD` composer.lock and composer.json files to container working directory
- docker-compose builds Composer image
- Composer container runs an install command to install plugins in Wordpress container

How do I do that?

I'm under the impression that with using containers we are not installing anything else in the Vagrant host directory.

**Comment by robmazan** [comment]
Ok, so think of containers as applications. When you are running a container you are actually running an application, that has all it's dependencies included, so you don't need to care about those. In this case, we are treating the Composer container as such an application. You don't have to install Composer directly and all of it's dependencies, rather just run it in a given folder, using it's containerized version.

When you are passing a volume from your host (which is the VM ran by Vagrant in our case), you are actually mounting a piece of the host file system to the container's file system. So if the container writes to that folder, that change will appear on the host (again, it's the Vagrant VM in our case, but if that volume happens to be under `/vagrant` on the Vagrant VM, VirtualBox will sync those changes back to your actual host (Mac, PC whatever)).

**Comment by hunterlester** [comment]
@robmazan Is it possible to run a container application, such as Composer, within another container's working directory? In our case, is it possible to run the Composer container application in the WordPress container application's working directory at `/var/www/html`?

**Comment by robmazan** [comment]
@hunterlester  It is possible, but with the method I've suggested you won't need that (I think). I mean you have the `/vagrant/platform/wp-content` folder on your Vagrant VM. The `/vagrant/platform/wp-content/plugins` and `/vagrant/platform/wp-content/themes` folders are already passed as a volume for the WordPress container to be available under `/var/www/html/wp-content/plugins` and `/var/www/html/wp-content/plugins/themes` respectively.

If I get it right, the Composer should install dependencies to `wp-content/plugins`, but it also installs some stuff into `wp-content/vendor`, that's required by those plugins.

Now instead of sharing the `/var/www/html/wp-content` from the WordPress container to the Composer container, it seems to be easier to just take the `/vagrant/platform/wp-content` folder from the Vagrant VM, and pass that as a volume, as it is basically the same, but in case you can run composer independently, like even if the WordPress container is not running.

**Comment by robmazan** [comment]
Just finished to build the container, and I think I see your problem: `wp-content` inside `wp-content`. :)

Trying to figure out something now.

**Comment by robmazan** [comment]
I've created a PR to your branch: https://github.com/hunterlester/JSHomes-Platform/pull/1

With this setup composer seems to work to the right folders, WordPress container is still building, so dunno if it works properly or not yet.

**Comment by hunterlester** [comment]
@robmazan Thank you :-D
I'll test it out in a couple of hours. Have to get in on a meeting.

**Comment by robmazan** [comment]
Seems to be working for me. At least I see the plugins in wp-admin. :)

To make it possible re-run composer without recreating the Vagrant VM, might worth to put the `docker run --rm -v /vagrant/platform:/app composer/composer install` command to a script file, similar to [start](https://github.com/jshomes/JSHomes-Platform/blob/master/platform/start), and call that from the Vagrantfile.

Or maybe:

```sh
alias composer=docker run --rm -v /vagrant/platform:/app composer/composer
```

And you can use the containerized Composer as `composer` command in your vagrant VM. :)

**Comment by hunterlester** [comment]
@robmazan awesome awesome! Oh man, thank you for helping me. Please know that your time has not been wasted. I'm going to take these lessons here and achieve a solid understanding of Docker.

**Comment by robmazan** [comment]
@hunterlester you're very welcome! :)

**Comment by hunterlester** [comment]
@ericelliott @robmazan @mattma 
Good morning
Okay, this fork has been updated to run composer `install` command from docker-compose file.
Installing nicely.

Next step:
Determining proper version constraints for plugins.
Should versions be strict and we determine, plugin by plugin, if an update is stable to the wordpress core?

**Comment by ericelliott** [comment]
> Should versions be strict and we determine, plugin by plugin, if an update is stable to the wordpress core?

I'm not sure. How does WordPress get notifications up updated plugin versions? Can we tap into the same mechanism?

**Comment by ericelliott** [comment]
BTW, I'll test and merge this branch ASAP.. more work can be done in a future branch.

**Comment by hunterlester** [comment]
> I'm not sure. How does WordPress get notifications up updated plugin versions? Can we tap into the same mechanism?

Doing some research and talking to Wordpress hackers around the building about this...

**Comment by ericelliott** [comment]
I hate to be a bottleneck.. @robmazan or @mattma - if either of you have tested this and verified that it works as expected, feel free to click "merge." Otherwise, I'll get to it in a day or two...

**Comment by robmazan** [comment]
@ericelliott can't check this right now, have only a 32 bit machine here... last time I was checking, it was working though, but that was on a Win machine. Shouldn't make a difference, but would worth to check on a Mac too IMO.

**Comment by ericelliott** [comment]
@robmazan Thanks for trying. =)

**Comment by rshipp** [comment]
What still needs to be tested here? I don't have access to a Mac, but I can test on Linux if it helps.

**Comment by hunterlester** [comment]
@rshipp Good morning Ryan. Thank you for your patience and please excuse the delay in response.

This is a good opportunity to determine what needs to be tested.

- All plugins listed in composer.json install and are available for activation in Wordpress dashboard
- When new plugin version is available and the decision is made to update from dashboard, does the composer.lock file successfully update it's data?

**Comment by rshipp** [comment]
@hunterlester No problem, and good afternoon (?) to you! It's actually night here now, but I should be able to work on this tomorrow. :smile: 

The initial setup was a bit tricky for me to figure out, so to increase the bus factor a bit, here's what I've done so far:

    vagrant up
    vagrant ssh

From inside the vagrant machine:

    cd platform
    sudo apt-get install php5-cli
    curl -sS https://getcomposer.org/installer | php
    php composer.phar install
    cp .env.example .env
    vi .env
    ./start

This gives me 

    vagrant@ubuntu-14:~/platform$ docker ps -a
    CONTAINER ID        IMAGE                COMMAND                  CREATED             STATUS                     PORTS                            NAMES
    937d6f06d91b        platform_wordpress   "/entrypoint.sh apach"   3 minutes ago       Up 3 minutes               3306/tcp, 0.0.0.0:9888->80/tcp   platform_wordpress_1
    9dbc220946b9        mysql:5.7            "/entrypoint.sh mysql"   3 minutes ago       Up 3 minutes               3306/tcp                         platform_mysql_1
    18d38321727d        composer/composer    "composer --ansi inst"   3 minutes ago       Exited (0) 3 minutes ago                                    platform_composer_1

I'm not sure if this is intentional, but the MySQL config in `wp-config.php` is not set, so I initially got a 500 Error when trying to access Wordpress. This is easy to fix:

    vagrant@ubuntu-14:~/platform$ docker inspect --format '{{ .NetworkSettings.IPAddress }}' 9dbc220946b9
    172.17.0.3
    vagrant@ubuntu-14:~/platform$ docker exec -it 937d6f06d91b /bin/bash
    root@937d6f06d91b:/var/www/html# vim wp-config.php  # Fill in DB_* fields, using the IP of the mysql container as HOST

Now the Wordpress installation is available at `localhost:9888`. 

I'll play with Composer tomorrow when I have more time.

**Comment by ericelliott** [comment]
Thanks @rshipp -- very helpful!

**Comment by rshipp** [comment]
* [x] All plugins listed in composer.json install and are available for activation in Wordpress dashboard

spectacle p19436

* [ ] When new plugin version is available and the decision is made to update from dashboard, does the composer.lock file successfully update it's data?

spectacle c19436

Checking the permissions, I found this:

    root@937d6f06d91b:/var/www/html# stat wp-content/plugins/
      File: 'wp-content/plugins/'
      Size: 4096            Blocks: 8          IO Block: 4096   directory
    Device: 23h/35d Inode: 222         Links: 1
    Access: (0755/drwxr-xr-x)  Uid: ( 1000/ UNKNOWN)   Gid: ( 1000/ UNKNOWN)
    Access: 2016-01-14 20:08:52.019355172 +0000
    Modify: 2016-01-11 14:40:18.000000000 +0000
    Change: 2016-01-14 20:00:42.739399067 +0000
     Birth: -

The plugins folder is mounted as a volume in `docker-compose.yml`, so the `chown` command in the `Dockerfile` isn't run there. If anyone knows how to fix this, feel free to jump in. Otherwise I'll keep poking around and see if I can figure it out.

**Comment by hunterlester** [comment]
Awesome. Thank you. I'll see if I can tinker with it as well.

**Comment by rshipp** [comment]
It looks like chown is broken at the Vagrant level too. [This article](http://jeremykendall.net/2013/08/09/vagrant-synced-folders-permissions/) seems relevant.

**Comment by hunterlester** [comment]
@rshipp @ericelliott 
Some questions popped into my head last night.
There will be several separate deployments of this platform, correct? Each maintained by different users/developers?
Perhaps it's a good thing that write permissions are not granted for the application to be able to update plugins and the `composer.lock` file?
I'm wondering if it would be better for just one dev to determine that a plugin update is safe and to make the necessary changes to `.lock` so that the file maintains only a history of safe versions instead of a history of every update from each deployment?

Let me know what you two think. 

@rshipp Thank you for that article.

> @hunterlester We'll want to simplify as much as possible and update the Getting Started Guide.

@ericelliott Just saw this. For sure. :-)

**Comment by ericelliott** [comment]
> There will be several separate deployments of this platform, correct? Each maintained by different users/developers?

Yes, but they should still be able to update their own installations from the Admin UI. Central dependency sounds like a good thing until you consider the  single point of failure principle. ;)

**Comment by hunterlester** [comment]
@ericelliott Vagrant file permissions are now allowing plugins to be updated.

I've removed the `composer.lock file` and added it to `.gitignore` so that there is no longer a central source of plugin versions and each deployment of the app will independently manage it's own plugin updates.

With these changes nothing has to be updated in the [getting started guide](https://github.com/jshomes/JSHomes-Platform/tree/master/docs)

As you find additional plugins that you want to add or remove from the platform, ping me in an issue with the plugin and I'll change the `composer.json` with the right version.

**Comment by ericelliott** [comment]
Fantastic!
